org-code-javabuilder/playground/src/main/java/org/code/playground/PlaygroundMessageHandler.java

```diff
@@ -26,20 +33,42 @@ private PlaygroundMessageHandler() {
   PlaygroundMessageHandler(OutputAdapter outputAdapter) {
     this.messagesEnabled = true;
     this.outputAdapter = outputAdapter;
+    this.queuedMessages = new LinkedList<>();
+    this.sendInProgress = false;
   }
 
   public void sendMessage(PlaygroundMessage message) {
     // only send messages if playground has not ended. Otherwise throw a runtime exception.
     if (!this.messagesEnabled) {
       throw new PlaygroundRuntimeException(PlaygroundExceptionKeys.INVALID_MESSAGE);
     } else {
-      this.outputAdapter.sendMessage(message);
+      this.queuedMessages.add(message);
     }
   }
 
+  public void sendBatchedMessages() {
+    if (this.queuedMessages.isEmpty() || this.sendInProgress) {
+      return;
+    }
+    this.sendInProgress = true;
```

Comment:
Can `sendBatchedMessages` be called while `sendInProgress` is true? I think we run student code in a single thread and querying for input is a blocking operation, right?